File: backends/java/zappy-application/src/main/java/com/zappymart/application/ordering/FindOrder.java
package com.zappymart.application.ordering;

import com.zappymart.application.Visitor;
import com.zappymart.application.ports.OrderRepository;
import com.zappymart.domain.ordering.Order;

import java.util.Optional;

public final class FindOrder {

    private final OrderRepository orderRepository;

    public FindOrder(OrderRepository orderRepository) {
        this.orderRepository = orderRepository;
    }

    public Optional<Order> execute(Visitor visitor, String orderId) {
        if (!visitor.isSignedIn()) {
            return Optional.empty();
        }
        return orderRepository.byIdOfCustomer(orderId, visitor.customerId());
    }
}

File: backends/kotlin/zappy-application/src/main/kotlin/nl/zappymart/application/ordering/FindOrder.kt
package nl.zappymart.application.ordering

import nl.zappymart.application.Visitor
import nl.zappymart.application.ports.OrderRepository
import nl.zappymart.domain.ordering.Order

class FindOrder(private val orders: OrderRepository) {

    fun execute(visitor: Visitor, orderId: String): Order? {
        val customerId = visitor.customerId ?: return null
        return orders.findForCustomer(orderId, customerId)
    }
}
